text/0000-generalized-type-ascription.md

```diff
@@ -0,0 +1,1956 @@
+
+// 2) Looking at `x: <X as Trait>::Assoc` the compiler:
+//
+//    a) Verifies that `X: Trait` and that `Trait` has an associated type `Assoc`.
+//    b) Substitutes `?T` for `<X as Trait>::Assoc`.
+//
+// 3) Thus, `good_7 : for<X: Trait> fn(Wrapping<X::Assoc>) -> ()`.
+fn good_7<X: Trait>(Wrapping(x: <X as Trait>::Assoc)) {}
+
+struct Bar<'a>(&'a bool);
+
+// 2) The compiler sees `x: Bar<'_>` and infers that `?T = Bar<'?a>`.
+//
+// 3) Looking at `'Bar<'?a>`, the compiler:
+//    a) introduces an input lifetime parameter `'a`
+//    b) and substitutes `'?a` for it.
+//
+// 4) Thus, `good_8 : for<'a> fn(Wrapping<Bar<'a>>) -> ()`.
+fn good_8(Wrapping(x: Bar<'_>)) {}
+
+struct Quux<'a, 'b: 'a>(&'a Foo, Bar<'b>);
+
+// 1) Looking at `Quux(x, $pat)`, the compiler infers `Quux<'?a, '?b>`
+//    where `'?b: '?a`.
+//
+// 2) Looking at `'Quux<'?a, '?b>`, the compiler:
+//    a) introduces input lifetime parameters `'a` and `'b`.
+//    b) adds a bound `'b: 'a`.
+//    c) and substitutes `'?a` and `'?b` for `'a` and `'b`.
+//    d) infers that `$pat : Bar<'b>` must hold.
+//
+// 3) Looking at `$pat = Bar(y)`, the compiler checks that `Bar(y) : Bar<'b>`.
+//
+// 4) Thus, `good_9 : for<'a, 'b: 'a> fn(Quux<'a, 'b>) -> ()`.
+fn good_9(Quux(x, Bar(y))) {}
+
+struct Product<A, B>(A, B);
+
+// 1) Looking at `Product($pat_1, $pat_2)`, the compiler infers that the
+//    parameter must be of type `Product<?T, ?U>` and adds unification variables
+//    `?T` and `?U`.
+//
+// 2) Looking at `$pat_1 = x: &i32`, the compiler:
+//    a) infers that `?T = &'?a i32`.
+//    b) introduces an input lifetime parameter `'a`.
+//    c) substitutes `'?a` for `'a`.
+//
+// 3) Looking at `$pat_2 = y: &u32`, the compiler:
+//    a) infers that `?T = &'?b u32`.
+//    b) introduces an input lifetime parameter `'b`.
+//    c) substitutes `'?b` for `'b`.
+//
+// 4) Thus, `good_10 : for<'a, 'b> fn(Product<&'a i32, &'b u32>) -> ()`.
+fn good_10(Product(x: &i32, y: &u32)) {}
+
+struct Wibble<'a>(&'a i32, Bar<'a>);
+
+// 1) Looking at `Wibble($pat_1, $pat_2)`, the compiler infers that the
+//    parameter must be of type `Wibble<'?a>` and adds unification variable `'?a`.
+//    a) The compiler introduces an input lifetime parameter `'a`
+//       and substitutes `'?a` for `'a`.
+//    b) Infers that `$pat_1 : &'a i32`  must hold.
+//    c) Infers that `$pat_2 : Bar<'a>`  must hold.
+//
+// 2) Looking at `$pat_1 = x: &i32`, the compiler:
+//    a) infers that `x : &'?b i32` and adds unification variable `'?b`.
+//    b) checks that `x : &'a i32`.
+//    c) this entails that `'?b = 'a` and so `'?b` is substituted for `'a`.
+//
+// 3) Looking at `$pat_2 = x: Bar<'_>`, the compiler:
+//    a) infers that `x : Bar<'?c>` and adds unification variable `'?c`.
+//    b) checks that `x : Bar<'a>`.
+//    c) this entails that `'?c = 'a` and so `'?c` is substituted for `'a`.
+//
+// 4) Thus, `good_11 : for<'a> fn(Wibble<'a>) -> ()`.
+fn good_11(Wibble(x: &i32, y: Bar<'_>)) {}
```

Comment:
Wow, this section is awesome!

I think I agree with all of them except for `good_11`: I would expect that 
```rust
fn good_11(Wibble(x: &i32, y: Bar<'_>)) {}
```
is equivalent to
```rust
fn good_11(Wibble(x: &'_ i32, y: Bar<'_>)) {}
```

is equivalent to 
```rust
fn good_11<'a, 'b>(Wibble(x: &'a i32, y: Bar<'b>)) {}
```

Which is illegal since it, unlike `Wibble`, doesn't require that `'a == 'b`.

That's because the version not wrapped in `Wibble`:
```
fn rutabaga(x: &i32, y: Bar<'_>) {}
```

would have two lifetimes, so I think the wrapped version should too.

This helps local reasoning, as you don't need to look at the declaration of Wibble to know whether the two lifetimes are the same or different (since `Product` and `Wibble` cases look the same in the function header).

Reply:
Glad you think so :)

I agree that `good_11` is the tricky one. The rule you imposed certainly does improve local reasoning and the expectation induced by `rutabaga` does seem like it would happen to folks sometimes.

However, I do wonder if people care and think so much about lifetime parameters (if you are dealing with unsafe code where this is important you should quantify explicitly) that it matters in practice.

+ At the call site `good_11` won't be callable with something where given `&'a i32` and `Bar<'b>` you have that `'a != 'b` because you won't be able to construct such a `Wibble<'?>` in the first place.

+ What remains is then confusion at the definition site of `good_11` where the user may think that `'a != 'b`. However, when the user actually tries to make use of this mistaken belief, the compiler should be able to give the user the actual signature of the function and/or say that the lifetimes are really the same. Thus, the user can therefore "learn on the job".

In any case, it is also true that your proposed rule allows strictly less than my rule so if we use your rule and then change our minds we can switch. So we are forward compatible.

Finally and as an aside, there are two ways to make your changed `good_11` shorter:

```rust
// in-band lifetimes (assuming we don't ditch it)
fn good_11(Wibble(x: &'a i32, y: Bar<'a>)) {}

// The types are fully determined by `Wibble<'_>` itself.
fn good_11(Wibble(x, y)) {}
```

I'll go and add an unresolved question for now; I think this is something which is subtle and where we'll want to weigh the trade-offs and discuss them on the RFC. :)

Reply:
...and I thought of more stuff for these examples: match ergonomics.

I think the following would work?
```rust
fn good_13(Foo(x): &_) {}
```

Giving a type of `for<'a> fn(&'a Foo)` and, in the body, `x: &'a usize`.